lib/evaluate/characteristics.cc

```diff
@@ -98,28 +98,46 @@ std::optional<TypeAndShape> TypeAndShape::Characterize(
   }
 }
 
-bool TypeAndShape::IsCompatibleWith(
-    parser::ContextualMessages &messages, const TypeAndShape &that) const {
+std::optional<TypeAndShape> TypeAndShape::Characterize(
+    const Expr<SomeType> &expr, FoldingContext &context) {
+  if (const auto *symbol{UnwrapWholeSymbolDataRef(expr)}) {
+    if (const auto *object{
+            symbol->detailsIf<semantics::ObjectEntityDetails>()}) {
+      return Characterize(*object);
+    }
+  }
+  if (auto type{expr.GetType()}) {
+    if (auto shape{GetShape(context, expr)}) {
+      TypeAndShape result{*type, std::move(*shape)};
+      if (type->category() == TypeCategory::Character) {
+        if (const auto *chExpr{UnwrapExpr<Expr<SomeCharacter>>(expr)}) {
+          if (auto length{chExpr->LEN()}) {
+            result.set_LEN(Expr<SomeInteger>{std::move(*length)});
+          }
+        }
+      }
+      return result;
+    }
+  }
+  return std::nullopt;
+}
+
+bool TypeAndShape::IsCompatibleWith(parser::ContextualMessages &messages,
+    const TypeAndShape &that, const char *thisIs, const char *thatIs,
+    bool isElemental) const {
   const auto &len{that.LEN()};
   if (!type_.IsTypeCompatibleWith(that.type_)) {
     std::stringstream lenstr;
     if (len) {
       len->AsFortran(lenstr);
     }
-    messages.Say("Target type '%s' is not compatible with '%s'"_err_en_US,
-        that.type_.AsFortran(lenstr.str()), type_.AsFortran());
+    messages.Say(
+        "%1$s type '%2$s' is not compatible with %3$s type '%4$s'"_err_en_US,
```

Comment:
What does `$s` mean in the error format string?

Reply:
It selects the actual argument to be formatted.  In this case, where `1$` through `4$` are used, the effect is the same as if they had not been present.  I inserted them here in response to a comment about easing internationalization.